CorsixTH/Src/th_lua.h

```diff
@@ -463,4 +467,79 @@ void luaT_pushtablebool(lua_State* L, const char* k, bool v);
 
 void preload_lua_package(lua_State* L, const char* name, lua_CFunction fn);
 
+namespace {
+int luaT_get_userdata_classname(lua_State* L, int idx, const char** class_name,
+                                void** userdata) {
+  void* p = lua_touserdata(L, idx);
+  *userdata = p;
+  if (p == nullptr) {
+    return 0;
+  }
+  int added_stack = 0;
+  if (lua_getmetatable(L, idx) == 0) {
+    std::printf("Warn: No metatable for userdata\n");
+    return 0;
+  }
+  lua_getfield(L, -1, "__index");
+  added_stack += 2;  // metadata and field
+  if (lua_type(L, -1) != LUA_TTABLE) {
+    lua_pop(L, added_stack);
+    std::printf("Warn: No __index field method table for userdata\n");
+    return 0;
+  }
+
+  if (lua_getmetatable(L, -1) == 0) {
+    lua_pop(L, added_stack);
+    std::printf("Warn: No metatable for method table of userdata\n");
+    return 0;
+  }
+  lua_getfield(L, -1, "__class_name");
+  added_stack += 2;  // method metatable and class name field
+  if (lua_type(L, -1) != LUA_TSTRING) {
+    lua_pop(L, added_stack);
+    std::printf("Warn: No __class_name field for userdata\n");
+    return 0;
+  }
+  *class_name = lua_tostring(L, -1);
+
+  return added_stack;
+}
+}  // namespace
+
+// Find the appropriate derived class of base for a given
+// userdata object.
+//
+// th_lua_internal.h defines our c++ binding class metatables:
+// metatable: { __index = method_table, ... }
+// method_table: { ... }
+// method_table metatable: { __class_name = class_name }
+template <typename B, typename T1, typename T2>
```

Comment:
I'm not really a fan that this is limited to exactly two types, and extending it requires copying most of the code.

This was the reason I moved the class name lookup logic out of this function.

I think if we run into a situation where we want more than 2 types I will move to a lambda based solution instead.

i.e.

```cpp
luaT_touserdata_base(lua_State* L, int idx, std::unordered_map<const char*, std::function<B*(void*)>> cast_map) {...}
```